checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -204,6 +205,35 @@ private UBQualifier accountForLowerBoundAnnos(Node node, UBQualifier qual, Strin
         return qual;
     }
 
+    private void propagateToSubterm(
+            Node target, Node offset, UBQualifier qualifier, CFStore store, boolean plusNode) {
+        UBQualifier existingQual =
+                UBQualifier.createUBQualifier(
+                        atypeFactory
+                                .getAnnotatedType(target.getTree())
+                                .getAnnotationInHierarchy(atypeFactory.UNKNOWN));
+
+        UBQualifier newQual =
+                existingQual.glb(
+                        plusNode
+                                ? qualifier.plusOffset(offset, atypeFactory)
+                                : qualifier.minusOffset(offset, atypeFactory));
+
+        Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), target);
+        store.insertValue(rightRec, atypeFactory.convertUBQualifierToAnnotation(newQual));
+    }
+
+    private void propagateToSubterms(
+            BinaryOperationNode node, UBQualifier qualifier, CFStore store) {
+        boolean plusNode = node instanceof NumericalAdditionNode;
+        // Otherwise this should be a numerical subtraction node.
+
+        propagateToSubterm(
+                node.getRightOperand(), node.getLeftOperand(), qualifier, store, plusNode);
+        propagateToSubterm(
```

Comment:
I've updated the code so that it doesn't incorrectly type `i` after `if (1 - i < a.length) {`. But, this comes at a cost; in that block `1 - i` now also doesn't have a type, so there's a false positive.